waku/v2/protocol/metadata/waku_metadata.go

```diff
@@ -0,0 +1,243 @@
+package metadata
+
+import (
+	"context"
+	"errors"
+	"math"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	libp2pProtocol "github.com/libp2p/go-libp2p/core/protocol"
+	"github.com/libp2p/go-msgio/pbio"
+	"github.com/multiformats/go-multiaddr"
+	"github.com/waku-org/go-waku/logging"
+	"github.com/waku-org/go-waku/waku/v2/protocol"
+	"github.com/waku-org/go-waku/waku/v2/protocol/enr"
+	"github.com/waku-org/go-waku/waku/v2/protocol/metadata/pb"
+	"go.uber.org/zap"
+)
+
+// MetadataID_v1 is the current Waku Metadata protocol identifier
+const MetadataID_v1 = libp2pProtocol.ID("/vac/waku/metadata/1.0.0")
+
+// WakuMetadata is the implementation of the Waku Metadata protocol
+type WakuMetadata struct {
+	network.Notifiee
+
+	h         host.Host
+	ctx       context.Context
+	cancel    context.CancelFunc
+	clusterID uint16
+	localnode *enode.LocalNode
+
+	log *zap.Logger
+}
+
+// NewWakuMetadata returns a new instance of Waku Metadata struct
+// Takes an optional peermanager if WakuLightPush is being created along with WakuNode.
+// If using libp2p host, then pass peermanager as nil
+func NewWakuMetadata(clusterID uint16, localnode *enode.LocalNode, log *zap.Logger) *WakuMetadata {
+	m := new(WakuMetadata)
+	m.log = log.Named("metadata")
+	m.clusterID = clusterID
+	m.localnode = localnode
+
+	return m
+}
+
+// Sets the host to be able to mount or consume a protocol
+func (wakuM *WakuMetadata) SetHost(h host.Host) {
+	wakuM.h = h
+}
+
+// Start inits the metadata protocol
+func (wakuM *WakuMetadata) Start(ctx context.Context) error {
+	if wakuM.clusterID == 0 {
+		wakuM.log.Warn("no clusterID is specified. Protocol will not be initialized")
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+
+	wakuM.ctx = ctx
+	wakuM.cancel = cancel
+
+	wakuM.h.Network().Notify(wakuM)
```

Comment:
Does libp2p accept more than 1 connectionNotifier or does it override? 
If it overrides, then the connectedness and connectionGater notifyDisconnect may not work.

Reply:
it accepts multiple connection notifiers!